[Attention] MLA support for V1

chenyang78 opened the pull request:
This PR is co-authored with Lucas Wilkinson.

```
VLLM_USE_V1="1" lm_eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=True --task gsm8k --num_fewshot=5 --limit 100
...
vllm (pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=True), gen_kwargs: (None), limit: 100.0, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.66|±  |0.0476|
|     |       |strict-match    |     5|exact_match|↑  | 0.66|±  |0.0476|
```

Signed-off-by: Yang Chen <yangche@fb.com>

<!--- pyml disable-next-line no-emphasis-as-heading -->

Commit:
[Attention] MLA support for V1

This PR is co-authored with Lucas Wilkinson.

```
VLLM_USE_V1="1" lm_eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=True --task gsm8k --num_fewshot=5 --limit 100
...
vllm (pretrained=deepseek-ai/DeepSeek-V2-Lite-Chat,tensor_parallel_size=2,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=True), gen_kwargs: (None), limit: 100.0, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.66|±  |0.0476|
|     |       |strict-match    |     5|exact_match|↑  | 0.66|±  |0.0476|
```

Signed-off-by: Yang Chen <yangche@fb.com>

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1420,10 +1413,18 @@ def get_kv_cache_spec(self) -> KVCacheSpec:
+                # i.e.
+                #     kv_lora_rank + pe_dim == kv_cache.size(2)
+                head_size = self.head_size if self.model_config.use_mla \
+                                           else attn_module.head_size
```
we are not using head_size for anything in MLA so we could just update here:
https://github.com/vllm-project/vllm/blob/4a8cfc75516f3df2ead816c0270b63470eb8e4ee/vllm/model_executor/models/deepseek_v2.py#L425
to be `kv_lora_rank + pe_dim`

chenyang78 replied:
Good point. Fixed. Thanks.

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -92,6 +92,26 @@ def __init__(
+                              or self.model_config.is_attention_free)
+        if not needs_attn_backend:
+            raise NotImplementedError(
+                "Non-Attention backend is not supported by V1 GPUModelRunner.")
```
nit: won't this check naturally happen by `get_attn_backend` returning None? can we just raise an `NotImplementedError` after `get_attn_backend` and log the inputs to `get_attn_backend` in order to cover this case?

chenyang78 replied:
Fixed. Thanks!

LucasWilkinson commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -356,6 +363,50 @@ def remove_request(self, req_id: str) -> Optional[int]:
+            self._req_ids[i2], self._req_ids[i1] # noqa
+        self.req_output_token_ids[i1], self.req_output_token_ids[i2] =\
+            self.req_output_token_ids[i2], self.req_output_token_ids[i1]
+        if old_id_i1 is not None and old_id_i2 is not None:
```
what happens if one is None and not the other?

chenyang78 replied:
Good catch. I added an assert to ensure we would not run into silent failures.

LucasWilkinson commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -80,7 +80,14 @@ def __init__(
         self.num_tokens_no_spec = np.zeros(max_num_reqs, dtype=np.int32)
         self.num_prompt_tokens = np.zeros(max_num_reqs, dtype=np.int32)
-        self.num_computed_tokens_cpu = np.empty(max_num_reqs, dtype=np.int32)
+        self.num_computed_tokens_cpu_tensor = torch.zeros(
```
hmmm I see why you did this (to pin the memory), but it might actually just be faster to do MLA context operations for chunked prefill, i.e. 
```
        if self.chunked_prefill_enabled and self._num_prefills > 0 \
             and context_lens_tensor[self._num_decodes:].max() > 0:
             ....

             # align max_context_chunk to page_size by rounding down,
             # currently the `gather_cache` kernel cannot handle
             # `context_chunk_starts` that are not aligned to page_size
             max_context_chunk = round_down(max_context_chunk, self.page_size)
             assert max_context_chunk > 0
             num_chunks = cdiv(context_lens_tensor.max(), max_context_chunk)

             # if `max_context_chunk = 256`, `num_chunks = 3`, and
             #   `num_prefills_with_context = 4`, create a tensor that looks like
             #  [[0, 0, 0, 0], [256, 256, 256, 256], [512, 512, 512, 512]]
             context_chunk_starts = \
                 torch.arange(num_chunks, device=device, dtype=torch.int32) \
                 .unsqueeze(1).expand(-1, num_reqs) \
                 * max_context_chunk
             .....
             context_chunk_seq_tot = chunk_seq_lens.sum(dim=1).tolist()
             assert max(context_chunk_seq_tot) <= \
                 self.chunked_prefill_workspace_size
```
on the cpu using numpy and then asynchronously transfer the results to the GPU, thoughts @WoosukKwon ?

chenyang78 replied:
I keep the current one as it and hope to give @WoosukKwon a bit more time for sharing his feedback. Thanks.

LucasWilkinson replied:
Also FYI this needs to be added to `condense` (if we dont get rid of it), we may want to use swap for condense https://github.com/chenyang78/vllm/pull/1#discussion_r1966311602 but there are micro-optimizations in the condense swap code that leverage the fact that its swapping with empty

@WoosukKwon do you have any idea how critical these micro-optimizations are?

LucasWilkinson replied:
@chenyang78 does `num_computed_tokens_cpu_tensor` need to be handled in swap states?

chenyang78 replied:
One naive question - would the following code handle it?

```
        self.num_computed_tokens_cpu[i1], self.num_computed_tokens_cpu[i2] =\
            self.num_computed_tokens_cpu[i2], self.num_computed_tokens_cpu[i1]
```

where 

```
        self.num_computed_tokens_cpu = \
            self.num_computed_tokens_cpu_tensor.numpy()
```

I see a similar pattern for ```self.token_ids_cpu_tensor``` and ```self.token_ids_cpu```, where we only swap ```self.token_ids_cpu```. Thanks.

LucasWilkinson replied:
oh I have no idea :/ we may want to double check the underlying data I think, I kinda assumed they were all shallow copies and that `num_computed_tokens_cpu` and `num_computed_tokens_cpu_tensor` would be disconnected other than sharing the same underlying data ptr (and one wouldn't update the other when shallow copied)

LucasWilkinson replied:
"double check the underlying data" just with some prints so we can convince ourselves

LucasWilkinson replied:
oh and I think `num_computed_tokens_cpu_tensor` needs to be handled in condense

chenyang78 replied:
Looks like they are aliased:

```
>>> import torch
>>> import numpy as np
>>> t = torch.zeros((2,2), device="cpu", dtype=torch.int32,pin_memory=False)
>>> n = t.numpy()
>>> t
tensor([[0, 0],
        [0, 0]], dtype=torch.int32)
>>> n[0][0] = 1
>>> t
tensor([[1, 0],
        [0, 0]], dtype=torch.int32)
```

LucasWilkinson replied:
```
>>> import torch
>>> import numpy as np
>>> t = torch.zeros((2,2), device="cpu", dtype=torch.int32,pin_memory=False)
>>> n = t.numpy()
>>> t2 = torch.zeros((2,2), device="cpu", dtype=torch.int32,pin_memory=False)
>>> n2 = t2.numpy()
>>> n[0][0] = 1
>>> t
tensor([[1, 0],
        [0, 0]], dtype=torch.int32)
>>> n2 = n
>>> t2
tensor([[0, 0],
        [0, 0]], dtype=torch.int32)
```

LucasWilkinson replied:
the swap doesn't work :(

LucasWilkinson replied:
sorry I was kinda more referring to after the swap, i.e.:

t -> Creates a buffer at address 0x5800
n -> then points to the buffer at 0x5800
t2 -> Creates a buffer at address 0x5a00
n2 -> then points to the buffer at 0x5a00

any changes to n will initially be reflected in t since they share the same buffer, but if do `n2 = n` now n and t2 point to the same buffer and t and n2 point to the same buffer, so we get a mismatch

```
>>> hex(n.__array_interface__['data'][0])
'0x583012bf5800'
>>> hex(t.data_ptr())
'0x583012bf5800'
>>> hex(n2.__array_interface__['data'][0])
'0x583012bf5a00'
>>> hex(t.data_ptr())
'0x583012bf5800'
>>> hex(t2.data_ptr())
'0x583012bf5a00'
>>> n2 = n
>>> hex(n2.__array_interface__['data'][0])
'0x583012bf5800'
>>> hex(t2.data_ptr())
'0x583012bf5a00'
```
look at the last few digits

LucasWilkinson replied:
we should also swap t and t2

LucasWilkinson replied:
the its correct:
```
>>> t2 = t
>>> hex(t2.data_ptr())
'0x583012bf5800'
```

chenyang78 replied:
Thanks for doing the experiments, Lucas!

Hmm, your example seems to have some subtle different from our swapping case, where we swap date from the same source? Like the example below:

```
>>> t = torch.zeros((2,), device="cpu", dtype=torch.int32,pin_memory=False)
>>> n = t.numpy()
>>> n[0] = 1
>>> n[1] = 2
>>> t
tensor([1, 2], dtype=torch.int32)
>>> n[0], n[1] = n[1], n[0]
>>> t
tensor([2, 1], dtype=torch.int32)
```

I could be totally wrong though :)

LucasWilkinson replied:
this looks good to me then! thanks for clarifying

chenyang78 replied:
No worries at all. Thank you so much for helping fix all the issues!

LucasWilkinson reviewed (commented):
Thanks for all the hardwork! Overall looks in pretty good shape to me, left a couple of comments

Broadly I think we should do some long term planning on how we abstract attention backends. Ultimately Id like to get rid of MetadataBuilders since these slowly over time became a mess in V0, but thats out of scope for this PR. I think we should also potentially investigate different Attention layers for MLA and "normal" attention in V1, also out of scope for this PR.

LucasWilkinson commented on vllm/v1/attention/backends/mla/common.py:
```diff
@@ -0,0 +1,1013 @@
+    # New for MLA (compared to FlashAttention)
+    # For chunked prefill
+    num_decodes: Optional[int] = None
+    num_decodes_tokens: Optional[int] = None
```
nit: rename to `num_decode_tokens`

Commit:
address review feedback

also included the fix from Lucas to resolve some cudagraph for
padded inputs / outputs

Signed-off-by: Yang Chen <yangche@fb.com>

chenyang78 commented:
> Thanks for all the hardwork! Overall looks in pretty good shape to me, left a couple of comments
> 
> Broadly I think we should do some long term planning on how we abstract attention backends. Ultimately Id like to get rid of MetadataBuilders since these slowly over time became a mess in V0, but thats out of scope for this PR. I think we should also potentially investigate different Attention layers for MLA and "normal" attention in V1, also out of scope for this PR.

Thanks for the prompt review! Yeah, I think it makes sense. I can follow-up with you later and figure our how to have a better design in this context.

chenyang78 commented on vllm/attention/layer.py:
```diff
@@ -166,15 +167,19 @@ def forward(
+            # We skip reshaping query, key and value tensors for the MLA
+            # backend since these tensors have different semantics and are
+            # processed differently.
+            if not self.use_mla:
```
It sounds a good idea to me. I can follow up on this.

chenyang78 commented on vllm/v1/attention/backends/triton_mla.py:
```diff
@@ -0,0 +1,110 @@
+
+    @staticmethod
+    def get_name() -> str:
+        return "TRITON_MLA_VLLM_V1"
```
Added. Thanks!

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1403,7 +1397,7 @@ def get_kv_cache_spec(self) -> KVCacheSpec:
                     block_size=block_size,
                     num_kv_heads=attn_module.num_kv_heads,
-                    head_size=attn_module.head_size,
+                    head_size=self.head_size,
```
with this updated now: https://github.com/vllm-project/vllm/pull/13789#discussion_r1968958809

I think this can be untouched? i.e. `attn_module.head_size`

chenyang78 replied:
Ah, nice catch! Totally overlooked it. Fixed. Thanks.

Commit:
restore to use attn_module.head_size

Tags:
Signed-off-by: Yang Chen <yangche@fb.com>

Commit:
included more fixes from Lucas

found those issues during benchmarking deepseek-V3

Signed-off-by: Yang Chen <yangche@fb.com>

chenyang78 commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -356,6 +363,64 @@ def remove_request(self, req_id: str) -> Optional[int]:
+            self.repetition_penalties_cpu[i2], self.repetition_penalties_cpu[i1]
+        self.min_p_cpu[i1], self.min_p_cpu[i2] =\
+            self.min_p_cpu[i2], self.min_p_cpu[i1]
+        if i1 in self.generators and i2 in self.generators:
```
@LucasWilkinson Could you please take a look if these changes make sense? Thanks!

LucasWilkinson replied:
oh oof im not that familiar with the V1 scheduler, lets ask in the rearch channel

WoosukKwon replied:
I think we should do
```python
g1 = self.generators.get(i1)
g2 = self.generators.get(i2)

if g1 is not None:
    self.generators[i2] = g1
if g2 is not None:
   self.generators[i1] = g2
```

LucasWilkinson replied:
should we be doing a similar thing for:
```
         old_id_i1 = self._req_ids[i1]
         old_id_i2 = self._req_ids[i2]

         assert old_id_i1 is not None and old_id_i2 is not None
         self.req_id_to_index[old_id_i1], self.req_id_to_index[old_id_i2] =\
             self.req_id_to_index[old_id_i2],
```
?

chenyang78 replied:
I committed the change suggested by @WoosukKwon . 

@LucasWilkinson I didn't touch the part that you pointed it out since I am not sure if we should make such a change or not. Let's wait for more feedback from Woosuk. Thanks.

LucasWilkinson replied:
I dont think we need it, I was just confused and being silly haha

Commit:
addressed feedback from Woosuk Kwon

Signed-off-by: Yang Chen <yangche@fb.com>

LucasWilkinson reviewed (approved):
LGTM now, thanks for all the hard work! this is very exciting :)

WoosukKwon commented:
Let me finish my review tonight.

WoosukKwon reviewed (approved):
Lgtm. Thanks for the work!

I see some rooms for optimization and cleanup, but let's merge this PR first. I will work on it this weekend if I get a chance.

Orthogonally, could you please benchmark the latency (of a small batch size)? While I'm glad that V1 shows higher throughput, I'm still a bit worried that the small ops in the attention causes significant CPU overheads for small batch sizes.

chenyang78 commented:
> Lgtm. Thanks for the work!
> 
> I see some rooms for optimization and cleanup, but let's merge this PR first. I will work on it this weekend if I get a chance.
> 
> Orthogonally, could you please benchmark the latency (of a small batch size)? While I'm glad that V1 shows higher throughput, I'm still a bit worried that the small ops in the attention causes significant CPU overheads for small batch sizes.

Thanks for the review! Could you share a benchmark command for checking the perf for small batch size? Thanks!

LucasWilkinson commented:
I ran some small requests with low QPS https://docs.google.com/spreadsheets/d/1toxQVaA7UPhmY57kv08Wdq0xtIcU8oBbqb9RE3Wy2_E/edit?usp=sharing

We do hit suffer from an ITL drop compared to V0 🫤

Im happy to start trying to optimize this next week
